File: MemorizationGame/app/src/main/java/com/example/memorizationgame/UI/Login/CreateAccountActivity.java
package com.example.memorizationgame.UI.Login;

import androidx.appcompat.app.AppCompatActivity;

import android.app.AlertDialog;
import android.content.DialogInterface;
import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageButton;
import android.widget.Toast;

import com.example.memorizationgame.Business.MusicService;
import com.example.memorizationgame.Business.UerAccount;
import com.example.memorizationgame.R;
import com.example.memorizationgame.UI.Account.AccountActivity;

public class CreateAccountActivity extends AppCompatActivity {



    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_create_account);


        //setting music
        final Intent serviceIntent = new Intent(CreateAccountActivity.this, MusicService.class);
        ImageButton musicPlayer = (ImageButton)findViewById(R.id.sound);
        if(MusicService.isplay == false){
            musicPlayer.setImageResource(R.drawable.start);
        }
        musicPlayer.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                if(MusicService.isplay == false){
                    startService(serviceIntent);
                    musicPlayer.setImageResource(R.drawable.stop);
                }else {
                    stopService(serviceIntent);
                    musicPlayer.setImageResource(R.drawable.start);
                }
            }
        });


        //setting new account button
        Button newAccount = (Button)findViewById(R.id.signup);
        newAccount.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                String name = ((EditText) findViewById(R.id.username)).getText().toString();
                String password = ((EditText) findViewById(R.id.password)).getText().toString();
                String re_password = ((EditText) findViewById(R.id.reenterPassword)).getText().toString();
                if("".equals(name)){
                    Toast.makeText(CreateAccountActivity.this, "The username could not be empty", Toast.LENGTH_SHORT).show();
                }else if("".equals(password)){
                    Toast.makeText(CreateAccountActivity.this, "The password could not be empty", Toast.LENGTH_SHORT).show();

                }else if("".equals(re_password)){
                    Toast.makeText(CreateAccountActivity.this, "The re-enter password could not be empty", Toast.LENGTH_SHORT).show();
                }else{
                    if(!password.equals(re_password)){
                        Toast.makeText(CreateAccountActivity.this, "The re-enter password is not correct,please re-enter again", Toast.LENGTH_SHORT).show();
                    }else {

                        //go to a new activity
                        Intent intent = new Intent(CreateAccountActivity.this, AccountActivity.class);
                        UerAccount user = new UerAccount(name,password);
                        intent.putExtra("user",user);
                        startActivity(intent);
                    }
                }


            }
        });

        //setting cancel button
        Button cancel = (Button)findViewById(R.id.cancel);
        cancel.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                finish();
            }
        });

        //setting question mark
        ImageButton questionmark = (ImageButton) findViewById(R.id.questionmark3);
        questionmark.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                showDialog(view );
            }
        });

    }

    public  void showDialog(View view){
        android.app.AlertDialog.Builder builder = new android.app.AlertDialog.Builder(this);
        builder.setTitle("Instruction");
        builder.setMessage("choose the right shape that you have already seen");
        builder.setPositiveButton("ok", new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialogInterface, int i) {

            }
        });
        AlertDialog dialog = builder.create();
        dialog.show();
    }


}

File: MemorizationGame/app/src/main/java/com/example/memorizationgame/UI/Login/CreateAccountActivity.kt
package com.example.memorizationgame.ui.login

import android.content.Intent
import android.os.Bundle
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.example.memorizationgame.R
import com.example.memorizationgame.Business.MusicService
import com.example.memorizationgame.Business.UerAccount
import com.example.memorizationgame.Data.UserRepository
import com.example.memorizationgame.databinding.ActivityCreateAccountBinding
import com.example.memorizationgame.UI.Account.AccountActivity
import com.example.memorizationgame.UI.Login.LoginViewModel

class CreateAccountActivity() : AppCompatActivity() {
    private val loginViewModel: LoginViewModel = LoginViewModel(UserRepository())
    private lateinit var binding: ActivityCreateAccountBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityCreateAccountBinding.inflate(layoutInflater)
        setContentView(binding.root)

        setupMusicService()
        setupOnClickListener()
        observeLiveData()
    }

    private fun setupMusicService(){
        val serviceIntent = Intent(this, MusicService::class.java)
        with(binding){
            if (!MusicService.isplay) {
                sound.setImageResource(R.drawable.start)
            }
            sound.setOnClickListener {
                if (!MusicService.isplay) {
                    startService(serviceIntent)
                    sound.setImageResource(R.drawable.stop)
                } else {
                    stopService(serviceIntent)
                    sound.setImageResource(R.drawable.start)
                }
            }
        }

    }

    private fun setupOnClickListener(){
        with(binding){
            // Setting new account button
            signup.setOnClickListener {
                val name = username.text.toString()
                val password = password.text.toString()
                val rePassword = reenterPassword.text.toString()
                when {
                    name.isEmpty() -> Toast.makeText(this@CreateAccountActivity,
                        "The username could not be empty", Toast.LENGTH_SHORT).show()
                    password.isEmpty() -> Toast.makeText(this@CreateAccountActivity,
                        "The password could not be empty", Toast.LENGTH_SHORT).show()
                    rePassword.isEmpty() -> Toast.makeText(this@CreateAccountActivity,
                        "The re-enter password could not be empty", Toast.LENGTH_SHORT).show()
                    password != rePassword -> Toast.makeText(this@CreateAccountActivity,
                        "The re-enter password is not correct, please re-enter again",
                        Toast.LENGTH_SHORT).show()
                    else -> {
                        Toast.makeText(this@CreateAccountActivity, "Loading...",
                            Toast.LENGTH_SHORT).show()
                        loginViewModel.createUser(UerAccount(name,password))
                    }

                }
            }

            // Setting cancel button
            cancel.setOnClickListener { finish() }

            // Setting question mark
            questionmark3.setOnClickListener { showDialog() }
        }
    }

    private fun showDialog() {
        val builder = android.app.AlertDialog.Builder(this)
        builder.setTitle("Instruction")
            .setMessage("Choose the right shape that you have already seen")
            .setPositiveButton("ok") { dialog, _ -> dialog.dismiss() }
            .create()
            .show()
    }

    private fun observeLiveData(){
        loginViewModel.user.observe(this,{
            val intent = Intent(this@CreateAccountActivity, AccountActivity::class.java)
            intent.putExtra("user", it)
            startActivity(intent)
        })

        loginViewModel.error.observe(this,{
            Toast.makeText(this, it, Toast.LENGTH_SHORT).show()
        })
    }
}
